setup.cfg

```
@@ -66,13 +64,16 @@ testing =
 	filelock>=3.4.0
 	ini2toml[lite]>=0.9
 	tomli-w>=1.0.0
+	tomli
 	pytest-timeout
 	pytest-perf; \
 		# workaround for jaraco/inflect#195, pydantic/pydantic-core#773 (see #3986)
 		sys_platform != "cygwin"
 	# for tools/finalize.py
 	jaraco.develop >= 7.21; python_version >= "3.9" and sys_platform != "cygwin"
 	pytest-home >= 0.5
+	# No Python 3.12 dependencies require importlib_metadata, but needed for type-checking since we import it directly
+	importlib_metadata; python_version >= "3.12"
```

You mean in this line: https://github.com/pypa/setuptools/blob/v69.1.1/setuptools/_importlib.py#L12?

Is there any other place we import it directly?

---

I do believe that's the only place.
Others indirect or similar imports include `from setuptools.extern import importlib_metadata` found a few lines lower
And this one in code template https://github.com/pypa/setuptools/blob/v69.1.1/setuptools/command/easy_install.py#L2141

The reason this line is needed for Python 3.12 is because for older versions `importlib_metadata` is a dependency of `build`, `twine` and `keyring` (itself a dependency of `jaraco.develop`, `twine`). Which seemingly all dropped it in 3.12.

You can't add a Python 3.12 check in code because users could have `importlib_metadata` on Python 3.12.
You can't type-ignore for a specific platform_version or os either.
Sensibly, the tests should install it to validate code that uses it.

The `; python_version >= "3.12"` part could be considered redundant.
```suggestion
	importlib_metadata
```

---

> Sensibly, the tests should install it to validate code that uses it.

This is a good point. By installing `importlib_metadata` we are exercising more code paths in `disable_importlib_metadata_finder` and improving coverage.